tmt/checks/watchdog.py

```diff
@@ -0,0 +1,470 @@
+                            command_output=ncat_output)
+
+        def _fail_connection_refused(ncat_output: str) -> None:
+            """ Handle failed connection """
+
+            logger.fail('connection refused')
+
+            guest_context.ssh_ping_failures += 1
+
+            report_progress(log,
+                            'ssh-ping',
+                            [
+                                '# connection refused',
+                                f'# failed {guest_context.ssh_ping_failures}'
+                                f' of {self.ssh_ping_threshold} allowed',
+                                ],
+                            command_output=ncat_output)
+
+        def _success(ncat_output: str) -> None:
+            """ Handle successfull response """
+
+            logger.verbose('Received successful response to SSH ping.', level=2)
+
+            report = [
+                '# successfull response'
+                ]
+
+            if guest_context.ssh_ping_failures != 0:
+                report.append(f'# replenished failure budget back to {self.ssh_ping_threshold}')
+
+            guest_context.ssh_ping_failures = 0
+
+            report_progress(
+                log,
+                'ssh-ping',
+                report,
+                command_output=ncat_output
+                )
+
+        try:
+            assert invocation.guest.primary_address is not None  # narrow type
+
+            output = tmt.utils.Command('nc',
+                                       '-zv',
+                                       invocation.guest.primary_address,
+                                       str(invocation.guest.port or 22)) .run(cwd=Path.cwd(),
+                                                                              logger=logger)
+
+            _success(output.stderr or '')
+
+        except tmt.utils.RunError as exc:
+            if exc.returncode == 1:
+                _fail_connection_refused(exc.stderr or '')
+
+            else:
+                _fail_unknown('\n'.join(render_run_exception_streams(exc.stdout, exc.stderr)))
+
+        logger.debug(
+            f'failed {guest_context.ssh_ping_failures}'
+            f' of {self.ssh_ping_threshold} allowed')
+
+        if guest_context.ssh_ping_failures >= self.ssh_ping_threshold:
+            logger.fail(f'exhausted {self.ssh_ping_threshold} SSH ping attempts')
+
+            self.notify(invocation, logger)
+
+
+@provides_check('watchdog')
+class Watchdog(CheckPlugin[WatchdogCheck]):
+    """
+    Take various actions when guest becomes unresponsive.
+
+    Watchdog runs selected probes every now and then, and when a given
+    number of `probes` fail, watchdog would run one or more of the
+    predefined `actions`.
+
+    Check comes with two probes, "ping" and "SSH ping", and single
+    action, "reboot".
+
+    * "ping" uses the classic ICMP echo to check whether the guest is
+      still up and running,
+    * "SSH ping" tries to establish SSH connection,
```

Comment:
How exactly should the `SSH ping` spelled in the config? `ssh-ping`? Would be probably good to include in the examples as well.

Reply:
Example added in 410efea7b260e7f0fced4e99b3609a0b67881e03.

But, well, yes, that's correct, and there's a `ssh-ping` key whose documentation points it out:

![2024-03-08_13-19](https://github.com/teemtee/tmt/assets/116509/043470d5-71c3-4ed3-8016-d21126cfd801)